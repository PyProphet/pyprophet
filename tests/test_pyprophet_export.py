from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
