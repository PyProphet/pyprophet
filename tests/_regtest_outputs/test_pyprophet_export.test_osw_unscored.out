    Charge  Intensity       RT  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_RT   assay_rt  decoy  delta_RT  delta_rt                         filename                   id  leftWidth        mz  rightWidth               run_id  transition_group_id  var_ms1_isotope_correlation_score  var_ms1_isotope_overlap_score  var_ms1_massdev_score  var_ms1_xcorr_coelution  var_ms1_xcorr_shape  var_ms2_bseries_score  var_ms2_dotprod_score  var_ms2_elution_model_fit_score  var_ms2_intensity_score  var_ms2_isotope_correlation_score  var_ms2_isotope_overlap_score  var_ms2_library_corr  var_ms2_library_dotprod  var_ms2_library_manhattan  var_ms2_library_rmsd  var_ms2_library_rootmeansquare  var_ms2_library_sangle  var_ms2_log_sn_score  var_ms2_manhattan_score  var_ms2_massdev_score  var_ms2_massdev_score_weighted  var_ms2_norm_rt_score  var_ms2_sonar_lag  var_ms2_sonar_log_diff  var_ms2_sonar_log_sn  var_ms2_sonar_log_trend  var_ms2_sonar_rsq  var_ms2_sonar_shape  var_ms2_xcorr_coelution  var_ms2_xcorr_coelution_weighted  var_ms2_xcorr_shape  var_ms2_xcorr_shape_weighted  var_ms2_yseries_score
0        2   207283.0  2661.55          117220.7482             854645.0      26.5  2595.5788      0    1.9379   65.9712  napedro_L120420_010_SW.mzXML.gz -4409520928686189639  2640.5100  728.8795   2705.3701 -8670811102654834151                    0                             0.9835                         0.1247                 1.3707                   0.0000               0.9907                    9.0                 0.7708                              NaN                   0.7811                             0.9962                         0.0000                0.9987                   0.9978                     0.0659                0.0239                          0.0262                  0.0725                4.7388                   0.7451                 0.3398                          0.1793                 0.0194                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   0.0000                            0.0000               0.9936                        0.9958                   11.0
1        2     6385.0  2605.74            8790.7812             104006.0      26.5  2595.5733      0    0.3198   10.1667  napedro_L120420_010_SW.mzXML.gz   260819276075322832  2575.6399  728.8795   2623.4399 -8670811102654834151                    0                             0.9555                         0.2667                 5.4202                   5.1430               0.6532                    2.0                 0.7610                              NaN                   0.0241                             0.9216                         0.1104                0.8271                   0.9764                     0.2223                0.0995                          0.1102                  0.3579                1.3130                   0.7675                 4.5391                          3.5103                 0.0032                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   7.0474                            2.3104               0.7806                        0.8341                    6.0
2        2     3838.0  2708.53            5750.4716              73215.2      26.5  2595.5750      0    3.3002  112.9550  napedro_L120420_010_SW.mzXML.gz  8534214264242363560  2705.3701  728.8795   2736.0901 -8670811102654834151                    0                            -0.3692                         0.7498                 7.1610                   6.7500               0.4827                    3.0                 0.8181                              NaN                   0.0145                             0.7660                         0.1334                0.8344                   0.9736                     0.2367                0.1055                          0.1166                  0.3772                0.6034                   0.6468                 2.5636                          1.1471                 0.0330                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   3.4656                            0.9347               0.6790                        0.7379                    5.0
3        2     5180.0  2832.77           10419.7435             241873.0      26.5  2595.5778      0    6.9026  237.1922  napedro_L120420_010_SW.mzXML.gz  5163914660633416481  2811.2000  728.8795   2855.5801 -8670811102654834151                    0                             0.6123                         0.4707                 8.9907                   4.0083               0.5985                    2.0                 0.7923                              NaN                   0.0195                             0.8418                         0.0911                0.9916                   0.9960                     0.0958                0.0387                          0.0426                  0.1243                0.6699                   0.6863                 4.7328                          2.9948                 0.0690                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   4.3568                            2.0950               0.6909                        0.6974                    6.0
4        2     2693.0  2795.06            4036.5600              25862.3      26.5  2595.5754      0    5.8092  199.4846  napedro_L120420_010_SW.mzXML.gz  6932937885234622359  2790.7200  728.8795   2811.2000 -8670811102654834151                    0                             0.1872                         2.4435                 1.8505                   4.0083               0.6422                    4.0                 0.7883                              NaN                   0.0101                             0.6804                         0.1794                0.4554                   0.9481                     0.3084                0.1494                          0.1882                  0.6202                0.6284                   0.6986                 5.4811                          3.8885                 0.0581                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   1.6487                            0.9186               0.7955                        0.7971                    6.0
..     ...        ...      ...                  ...                  ...       ...        ...    ...       ...       ...                              ...                  ...        ...       ...         ...                  ...                  ...                                ...                            ...                    ...                      ...                  ...                    ...                    ...                              ...                      ...                                ...                            ...                   ...                      ...                        ...                   ...                             ...                     ...                   ...                      ...                    ...                             ...                    ...                ...                     ...                   ...                      ...                ...                  ...                      ...                               ...                  ...                           ...                    ...
95       2   597887.0  2230.18          269150.5777            1192530.0      16.3  2245.2318      0   -0.3700  -15.0518  napedro_L120420_010_SW.mzXML.gz  6870255268859409918  2213.2600  612.3184   2247.3999 -8670811102654834151                   19                             0.9939                         0.1300                 2.0203                   0.0000               0.9990                    8.0                 0.7684                              NaN                   0.6626                             0.9928                         0.0000                0.8658                   0.9884                     0.1569                0.0719                          0.0725                  0.2613                4.7826                   0.7659                 1.6355                          1.5429                 0.0037                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   0.0000                            0.0000               0.9992                        0.9993                    8.0
96       2    48058.0  2291.53            6480.8607              75465.1      16.3  2245.2345      0    1.4088   46.2955  napedro_L120420_010_SW.mzXML.gz  2043199813358518344  2267.8799  612.3184   2298.6101 -8670811102654834151                   19                             0.7397                         3.0841                 3.4962                   5.3116               0.4407                    5.0                 0.6916                              NaN                   0.0533                             0.7474                         0.0866               -0.3525                   0.8138                     0.5478                0.2791                          0.3394                  1.0605                1.2194                   0.8070                 2.2562                          2.3997                 0.0141                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   4.4101                            1.9926               0.7236                        0.7141                    7.0
97       2    16553.0  2317.38            9656.7598              89588.0      16.3  2245.2345      0    2.1584   72.1455  napedro_L120420_010_SW.mzXML.gz  6262215160571261022  2302.0200  612.3184   2332.7400 -8670811102654834151                   19                             0.4212                         0.8821                 2.1071                   7.7080               0.4296                    4.0                 0.7153                              NaN                   0.0183                             0.0829                         0.2014               -0.0107                   0.9622                     0.2794                0.1337                          0.1370                  0.5120                0.2548                   0.7596                 1.9435                          1.6447                 0.0216                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   0.7830                            0.1896               0.7600                        0.7459                    7.0
98       2    20746.0  2120.97           10568.0806              85676.6      16.3  2245.2373      0   -3.5368 -124.2673  napedro_L120420_010_SW.mzXML.gz  -610141049182829192  2100.6001  612.3184   2141.5701 -8670811102654834151                   19                             0.5016                         0.9312                12.3692                   1.8165               0.6723                    4.0                 0.7176                              NaN                   0.0230                             0.6551                         0.4036                0.5606                   0.9754                     0.2021                0.0921                          0.1010                  0.3719                0.7732                   0.7428                 4.7941                          2.9011                 0.0354                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   2.5491                            1.1289               0.8015                        0.7928                    5.0
99       2    10959.0  2538.15           10810.2698              56553.9      16.3  2245.2307      0    8.5599  292.9193  napedro_L120420_010_SW.mzXML.gz -5430403952310232561  2520.5100  612.3184   2544.4099 -8670811102654834151                   19                             0.7016                         0.5650                 8.3246                   1.8165               0.5954                    2.0                 0.6942                              NaN                   0.0121                             0.4590                         0.6792               -0.1524                   0.9216                     0.3502                0.1629                          0.1915                  0.6979                0.6856                   0.7628                 4.0046                          4.4541                 0.0856                NaN                     NaN                   NaN                      NaN                NaN                  NaN                   1.4718                            0.6660               0.7586                        0.6855                    6.0

[100 rows x 50 columns]
